### ISSUE TITLE
Redesign of ExoPlayer layout and some new options for Android

### ISSUE THREAD

**Issue by PhantomPainX** [jepiqueau/capacitor-video-player#104]
I redesigned the player and made it more customizable with some new options:

- title (string): the title of the video.
- smallTitle (string): the subtitle below the title.
- mainColor (string): exo progress bar color (played and scrubber color) | default: white
- spinnerColor (string): android default indeterminated color while the video is loading. | default: white

### Result

![Screenshot_2022-08-23-15-42-20-59_21ae4d291e4b2391d104fa7a6e40562e](https://user-images.githubusercontent.com/47803967/186252993-275c9dd2-6e72-436e-934c-3dce8ee35c0c.jpg)

Made this using this tutorial https://levelup.gitconnected.com/customize-exoplayer-overlay-look-like-youtube-player-14fdd6d4583d

Let me know if you want a PR of this, I think this design is better in UI and UX.

**Comment by PhantomPainX** [comment]
@jepiqueau I was trying to add support for Chromecast too but ExoPlayer needs to update to the version 2.16.0 at least due to this:

`Upgrade the WorkManager extension to depend on androidx.work:work-runtime:2.7.0. Earlier versions of work-runtime are not compatible with apps targeting Android 12, and will crash with an IllegalArgumentException when creating PendingIntents (https://github.com/google/ExoPlayer/issues/9181).`

Basically I could connect successfully to my chromecast but the app crash due to that error. If you could upgrade this it would very nice :)

**Comment by jepiqueau** [comment]
@PhantomPainX this is far too much work for the time i have free at the moment. I may have a look in the future but not a near one

**Comment by PhantomPainX** [comment]
> @PhantomPainX this is far too much work for the time i have free at the moment. I may have a look in the future but not a near one

Sure, no problem 👌

What about the design I made? Can I do a PR?

**Comment by jepiqueau** [comment]
@PhantomPainX i merge the new design. If you have time you can try making a PR for the  upgrade to Exoplayer 2.16.0 but it is quite a work.
My next task is to upgrade to Capacitor 4 when i have a free time

**Comment by PhantomPainX** [comment]
> @PhantomPainX i merge the new design. If you have time you can try making a PR for the  upgrade to Exoplayer 2.16.0 but it is quite a work.
> My next task is to upgrade to Capacitor 4 when i have a free time 

I'm on that now, I updated to 2.16.0 and added Cast support (works), nevertheless I'm having problems with subtitles and I need to configure the controls and local player when Cast mode is playing

**Comment by jepiqueau** [comment]
@PhantomPainX tell me about your progress i will wait for it as you have starter to upgrade to capacitor 4 thanks for helping me

**Comment by PhantomPainX** [comment]
> @PhantomPainX tell me about your progress i will wait for it as you have starter to upgrade to capacitor 4 thanks for helping me

Hey, I updated successfully to ExoPlayer v2.16.0 and added Chromecast support (while it's casting, I disabled all local player buttons. Making custom buttons for chromecast in the layout is a lot of work and I won't do it haha).

Plus I updated this function `private MediaSource getSubTitle(MediaSource mediaSource, Uri sturi, DataSource.Factory dataSourceFactory)` to support this version of ExoPlayer, but I'm unable to test it... If you can help me with that would be awesome, I will make a PR in some hours

**Comment by jepiqueau** [comment]
@PhantomPainX what are the reasons for which you cannot test the subtitles

**Comment by PhantomPainX** [comment]
> @PhantomPainX what are the reasons for which you cannot test the subtitles

I don't have an url right now, If you provide me a valid one I can test it now

**Comment by jepiqueau** [comment]
@PhantomPainX look at `https://github.com/jepiqueau/angular-videoplayer-app-starter/blob/master/src/app/home/home.page.ts` you will find video urls with subtitles

**Comment by PhantomPainX** [comment]
> @PhantomPainX look at `https://github.com/jepiqueau/angular-videoplayer-app-starter/blob/master/src/app/home/home.page.ts` you will find video urls with subtitles

Ok, I finally made subtitles to work

![Screenshot_2022-08-29-21-04-01-53_21ae4d291e4b2391d104fa7a6e40562e.jpg](https://user-images.githubusercontent.com/47803967/187325292-9e9d6bd1-fad8-40a0-b7ea-931924db5b2c.jpg)

I will make the PR, some methods are deprecated but stills works (I'd update them but I don't have more time)

**Comment by jepiqueau** [comment]
@PhantomPainX Can you enumerate the methods which are deprecated

**Comment by jepiqueau** [comment]
@PhantomPainX i make the merge, publish it on npm and try to run it on `angular-videoplayer-app-starter` and i got a crash

```
2022-08-31 11:57:29.759 14756-14756/com.jeep.app.ioniccapacitorvideoplayer E/AndroidRuntime: FATAL EXCEPTION: main
    Process: com.jeep.app.ioniccapacitorvideoplayer, PID: 14756
    java.lang.RuntimeException: Unable to start activity ComponentInfo{com.jeep.app.ioniccapacitorvideoplayer/com.jeep.app.ioniccapacitorvideoplayer.MainActivity}: java.lang.IllegalStateException: The fully qualified name of the implementation of OptionsProvider must be provided as a metadata in the AndroidManifest.xml with key com.google.android.gms.cast.framework.OPTIONS_PROVIDER_CLASS_NAME.
        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3635)
        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:3792)
        at android.app.servertransaction.LaunchActivityItem.execute(LaunchActivityItem.java:103)
        at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:135)
        at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:95)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2210)
        at android.os.Handler.dispatchMessage(Handler.java:106)
        at android.os.Looper.loopOnce(Looper.java:201)
        at android.os.Looper.loop(Looper.java:288)
        at android.app.ActivityThread.main(ActivityThread.java:7839)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:548)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1003)
     Caused by: java.lang.IllegalStateException: The fully qualified name of the implementation of OptionsProvider must be provided as a metadata in the AndroidManifest.xml with key com.google.android.gms.cast.framework.OPTIONS_PROVIDER_CLASS_NAME.
        at com.google.android.gms.cast.framework.CastContext.zzg(com.google.android.gms:play-services-cast-framework@@21.1.0:6)
        at com.google.android.gms.cast.framework.CastContext.getSharedInstance(com.google.android.gms:play-services-cast-framework@@21.1.0:4)
        at com.jeep.app.ioniccapacitorvideoplayer.MainActivity.onCreate(MainActivity.java:12)
        at android.app.Activity.performCreate(Activity.java:8051)
        at android.app.Activity.performCreate(Activity.java:8031)
        at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1329)
        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3608)
        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:3792) 
        at android.app.servertransaction.LaunchActivityItem.execute(LaunchActivityItem.java:103) 
        at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:135) 
        at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:95) 
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2210) 
        at android.os.Handler.dispatchMessage(Handler.java:106) 
        at android.os.Looper.loopOnce(Looper.java:201) 
        at android.os.Looper.loop(Looper.java:288) 
        at android.app.ActivityThread.main(ActivityThread.java:7839) 
        at java.lang.reflect.Method.invoke(Native Method) 
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:548) 
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1003) 
```
any idea ?is 
```
<?xml version="1.0" encoding="utf-8"?>
<manifest xmlns:android="http://schemas.android.com/apk/res/android"
    package="com.jeep.app.ioniccapacitorvideoplayer">

    <application
        android:allowBackup="true"
        android:icon="@mipmap/ic_launcher"
        android:label="@string/app_name"
        android:roundIcon="@mipmap/ic_launcher_round"
        android:supportsRtl="true"
        android:theme="@style/AppTheme">

        <activity
            android:configChanges="orientation|keyboardHidden|keyboard|screenSize|locale|smallestScreenSize|screenLayout|uiMode"
            android:name="com.jeep.app.ioniccapacitorvideoplayer.MainActivity"
            android:label="@string/title_activity_main"
            android:theme="@style/AppTheme.NoActionBarLaunch"
            android:launchMode="singleTask"
            android:exported="true"
            android:alwaysRetainTaskState="true"
            android:resizeableActivity="true"
            android:supportsPictureInPicture="true">


          <intent-filter>
              <action android:name="android.intent.action.MAIN" />
              <category android:name="android.intent.category.LAUNCHER" />
          </intent-filter>
          <meta-data
            android:name="com.google.android.gms.cast.framework.OPTIONS_PROVIDER_CLASS_NAME"
            android:value="com.google.android.exoplayer2.ext.cast.DefaultCastOptionsProvider" >
          </meta-data>

        </activity>

        <provider
            android:name="androidx.core.content.FileProvider"
            android:authorities="${applicationId}.fileprovider"
            android:exported="false"
            android:grantUriPermissions="true">
            <meta-data
                android:name="android.support.FILE_PROVIDER_PATHS"
                android:resource="@xml/file_paths">
           </meta-data>
        </provider>
    </application>

    <!-- Permissions -->

    <uses-permission android:name="android.permission.INTERNET" />
    <uses-permission android:name="android.permission.READ_EXTERNAL_STORAGE" />

</manifest>

```
the manifest file

**Comment by jepiqueau** [comment]
@PhantomPainX Can you share your testing app?

**Comment by PhantomPainX** [comment]
> @PhantomPainX i make the merge, publish it on npm and try to run it on `angular-videoplayer-app-starter` and i got a crash
> 
> ```
> 2022-08-31 11:57:29.759 14756-14756/com.jeep.app.ioniccapacitorvideoplayer E/AndroidRuntime: FATAL EXCEPTION: main
>     Process: com.jeep.app.ioniccapacitorvideoplayer, PID: 14756
>     java.lang.RuntimeException: Unable to start activity ComponentInfo{com.jeep.app.ioniccapacitorvideoplayer/com.jeep.app.ioniccapacitorvideoplayer.MainActivity}: java.lang.IllegalStateException: The fully qualified name of the implementation of OptionsProvider must be provided as a metadata in the AndroidManifest.xml with key com.google.android.gms.cast.framework.OPTIONS_PROVIDER_CLASS_NAME.
>         at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3635)
>         at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:3792)
>         at android.app.servertransaction.LaunchActivityItem.execute(LaunchActivityItem.java:103)
>         at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:135)
>         at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:95)
>         at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2210)
>         at android.os.Handler.dispatchMessage(Handler.java:106)
>         at android.os.Looper.loopOnce(Looper.java:201)
>         at android.os.Looper.loop(Looper.java:288)
>         at android.app.ActivityThread.main(ActivityThread.java:7839)
>         at java.lang.reflect.Method.invoke(Native Method)
>         at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:548)
>         at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1003)
>      Caused by: java.lang.IllegalStateException: The fully qualified name of the implementation of OptionsProvider must be provided as a metadata in the AndroidManifest.xml with key com.google.android.gms.cast.framework.OPTIONS_PROVIDER_CLASS_NAME.
>         at com.google.android.gms.cast.framework.CastContext.zzg(com.google.android.gms:play-services-cast-framework@@21.1.0:6)
>         at com.google.android.gms.cast.framework.CastContext.getSharedInstance(com.google.android.gms:play-services-cast-framework@@21.1.0:4)
>         at com.jeep.app.ioniccapacitorvideoplayer.MainActivity.onCreate(MainActivity.java:12)
>         at android.app.Activity.performCreate(Activity.java:8051)
>         at android.app.Activity.performCreate(Activity.java:8031)
>         at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1329)
>         at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3608)
>         at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:3792) 
>         at android.app.servertransaction.LaunchActivityItem.execute(LaunchActivityItem.java:103) 
>         at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:135) 
>         at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:95) 
>         at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2210) 
>         at android.os.Handler.dispatchMessage(Handler.java:106) 
>         at android.os.Looper.loopOnce(Looper.java:201) 
>         at android.os.Looper.loop(Looper.java:288) 
>         at android.app.ActivityThread.main(ActivityThread.java:7839) 
>         at java.lang.reflect.Method.invoke(Native Method) 
>         at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:548) 
>         at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1003) 
> ```
> 
> any idea ?is
> 
> ```
> <?xml version="1.0" encoding="utf-8"?>
> <manifest xmlns:android="http://schemas.android.com/apk/res/android"
>     package="com.jeep.app.ioniccapacitorvideoplayer">
> 
>     <application
>         android:allowBackup="true"
>         android:icon="@mipmap/ic_launcher"
>         android:label="@string/app_name"
>         android:roundIcon="@mipmap/ic_launcher_round"
>         android:supportsRtl="true"
>         android:theme="@style/AppTheme">
> 
>         <activity
>             android:configChanges="orientation|keyboardHidden|keyboard|screenSize|locale|smallestScreenSize|screenLayout|uiMode"
>             android:name="com.jeep.app.ioniccapacitorvideoplayer.MainActivity"
>             android:label="@string/title_activity_main"
>             android:theme="@style/AppTheme.NoActionBarLaunch"
>             android:launchMode="singleTask"
>             android:exported="true"
>             android:alwaysRetainTaskState="true"
>             android:resizeableActivity="true"
>             android:supportsPictureInPicture="true">
> 
> 
>           <intent-filter>
>               <action android:name="android.intent.action.MAIN" />
>               <category android:name="android.intent.category.LAUNCHER" />
>           </intent-filter>
>           <meta-data
>             android:name="com.google.android.gms.cast.framework.OPTIONS_PROVIDER_CLASS_NAME"
>             android:value="com.google.android.exoplayer2.ext.cast.DefaultCastOptionsProvider" >
>           </meta-data>
> 
>         </activity>
> 
>         <provider
>             android:name="androidx.core.content.FileProvider"
>             android:authorities="${applicationId}.fileprovider"
>             android:exported="false"
>             android:grantUriPermissions="true">
>             <meta-data
>                 android:name="android.support.FILE_PROVIDER_PATHS"
>                 android:resource="@xml/file_paths">
>            </meta-data>
>         </provider>
>     </application>
> 
>     
> 
>     <uses-permission android:name="android.permission.INTERNET" />
>     <uses-permission android:name="android.permission.READ_EXTERNAL_STORAGE" />
> 
> </manifest>
> ```
> 
> the manifest file

Your declared wrong this in your app manifest:

```
<meta-data
        android:name="com.google.android.gms.cast.framework.OPTIONS_PROVIDER_CLASS_NAME"
        android:value="com.google.android.exoplayer2.ext.cast.DefaultCastOptionsProvider" />
```

You have:

```
<meta-data
   android:name="com.google.android.gms.cast.framework.OPTIONS_PROVIDER_CLASS_NAME"
   android:value="com.google.android.exoplayer2.ext.cast.DefaultCastOptionsProvider" >
</meta-data>
```

**Comment by jepiqueau** [comment]
@PhantomPainX it is working but i do not see the PiP mode button anymore

**Comment by jepiqueau** [comment]
@PhantomPainX i have  still to test for the Chromecast also

**Comment by PhantomPainX** [comment]
> @PhantomPainX it is working but i do not see the PiP mode button anymore

make sure you have pip enabled, it works for me

**Comment by jepiqueau** [comment]
@PhantomPainX OK i will have a closer look

**Comment by jepiqueau** [comment]
@PhantomPainX Emulator Pixel XL API 31 when i press on the button which i believe is PIP i got the video enlarged which is not the PIP

**Comment by PhantomPainX** [comment]
> @PhantomPainX Emulator Pixel XL API 31 when i press on the button which i believe is PIP i got the video enlarged which is not the PIP

Ok, let me test in that phone

**Comment by PhantomPainX** [comment]
Ok, tested in my IONIC app with that emulator and it works. I'm gonna test it in your app now

<img width="323" alt="Captura de Pantalla 2022-08-31 a la(s) 10 46 51" src="https://user-images.githubusercontent.com/47803967/187708451-056c6140-9956-421c-9d96-9899668c5c03.png">
<img width="315" alt="Captura de Pantalla 2022-08-31 a la(s) 10 47 26" src="https://user-images.githubusercontent.com/47803967/187708477-e064f131-980c-4b76-a1f7-811abd9be25e.png">

**Comment by PhantomPainX** [comment]
> @PhantomPainX Sorry my mistake i add a PipEnable set to false in one part of my code Seems that everything work fine. Great work. i am please that you did it. I still have to test Chrome cast. I am going to publish a new release with the API.md modified for the manifest.xml. I will look later at upgrading it for Capacitor 4

Glad it worked, please include all steps I wrote in the PR to make the plugin to work, that include (app gradle, manifest and mainactivity).

There is a bug in the chromecast button color too, it will become black when dark mode is disabled, I fixed it yesterday but deleted it today after testing the main branch of the plugin :(. I will probably upload the fix today

**Comment by jepiqueau** [comment]
@PhantomPainX i still do not see the chromecast button. i made some modification to allow chromecast when you want to pick the videos from the gallery. I would like to be able to test chromecast prior to move to capacitor4

**Comment by PhantomPainX** [comment]
> @PhantomPainX i still do not see the chromecast button. i made some modification to allow chromecast when you want to pick the videos from the gallery. I would like to be able to test chromecast prior to move to capacitor4

I don't know if Chromecast would work with local videos in your device storage. From what I understand when you cast a video, you send the video url, metadata and etc... in a mediaitem to the cast device and that device is responsible to play the video externally of the phone.

What I mean is that the phone doesn't send the video directly to the cast device, the cast device itself play the video

At least this is what I understand how it works, idk if I'm wrong

**Comment by jepiqueau** [comment]
@PhantomPainX in fact i have a samsung TV which is chromecast enable and a physical device  Samsung Galaxy S20 FE and i cannot set it up which cast app are you using on your android device ?

**Comment by PhantomPainX** [comment]
> @PhantomPainX in fact i have a samsung TV which is chromecast enable and a physical device  Samsung Galaxy S20 FE and i cannot set it up which cast app are you using on your android device ? 

I don't know what you mean with a Cast app, I'm using the plugin with IONIC

Are you trying to cast a remote video or internal?

**Comment by jepiqueau** [comment]
@PhantomPainX can you share your testing app so i can understand how to use it

**Comment by PhantomPainX** [comment]
> @PhantomPainX can you share your testing app so i can understand how to use it

I can't, sorry. It is an app for a client and I can't share it.

The most I can give you is the code where I initializate the player and some video urls I'm using

**Comment by jepiqueau** [comment]
@PhantomPainX yes please do so

**Comment by PhantomPainX** [comment]
> @PhantomPainX yes please do so

Ok, i had to edit the code because i had to many services, classes, etc. Anyways i didn't change the methods i use
```

import { Component, OnInit } from '@angular/core';

import { CapacitorVideoPlayer } from 'capacitor-video-player';

import { HTTP } from '@awesome-cordova-plugins/http/ngx';

@Component({
  selector: 'app-videos',
  templateUrl: './videos.page.html',
  styleUrls: ['./videos.page.scss'],
})
export class VideosPage implements OnInit {


  private capacitorVideoPlayer: any = CapacitorVideoPlayer;

  private testEmbedUrls: any[string] = [
    "https://fembed.com/v/7zv-jnkg2ox",
    "https://fembed.com/v/3q91j75d29y",
    "https://fembed.com/v/4dvjxzd0y91"
  ]

  private embed_url = "https://fembed.com/v/7zv-jnkg2ox";
  
  constructor(private http: HTTP) { }

  ngOnInit() {
    this.playVideo(this.testEmbedUrls[0]);
  }

  async playVideo(embedUrl: string) {

    this.fembedResolver(embedUrl).then((videos) => {
      this.nativePlayer2(videos[0], "Title", "smallTitle");
    });
  }

  // With this you get raw videos
  async fembedResolver(url: string) {
    url = url.replace("fembed.com", "vanfem.com");
    url = url.replace("/v/", "/api/source/");

    return new Promise((resolve, reject) => {

      this.http.post(url, {}, {
      'Content-Type': 'application/json',
      'Referer': url
      }).then(data => {
        console.log(data.data);
        const json = JSON.parse(data.data);
        const mVideos = json.data;
        var videos = [];

        mVideos.forEach(video => {
          videos.push({
            'label': video.label,
            'file': video.file,
            'referer': url
          });
        });
        const document = videos;
        resolve(document);
      }).catch(err => {
        reject(err); 
      });
    });
  }

  async nativePlayer2(video: any, title, smallTitle) {

    console.log("URL: ",encodeURI(video.file));
    await this.capacitorVideoPlayer.initPlayer({
      mode: 'fullscreen',
      url: encodeURI(video.file),
      playerId: 'player1',
      headers: {
        Referer: video.referer
      },
      title: title,
      smallTitle: smallTitle,
      accentColor: "#f0b400",
      chromecast: true
    }).then(() => {
      console.log('Player initialized');
      this.capacitorVideoPlayer.play({playerId: 'player1'}).then(() => {
        //this.screenOrientation.lock(this.screenOrientation.ORIENTATIONS.ANY);
        console.log('Player playing');

      }).catch(err => {
        console.log('Error playing', err);
        this.capacitorVideoPlayer.stopAllPlayers().then(() => {
          console.log('All players stopped');
        });
      });
    }).catch(err => {
      console.log('Error initializing player', err);
      this.capacitorVideoPlayer.stopAllPlayers().then(() => {
        console.log('All players stopped');
      });
    });
    
    await this.capacitorVideoPlayer.addListener('jeepCapVideoPlayerExit', (data) => {
      console.log('Player exited');
      //this.screenOrientation.lock(this.screenOrientation.ORIENTATIONS.PORTRAIT);
    });

    await this.capacitorVideoPlayer.addListener('jeepCapVideoPlayerEnded', (data) => {
      console.log('Player ended');
      //this.screenOrientation.lock(this.screenOrientation.ORIENTATIONS.PORTRAIT);
    });
  }

}

```

**Comment by PhantomPainX** [comment]
@jepiqueau I'm having issues with some video links, some throws 403 source error (In cordova-plugin-streaming-media with MediaPlayer works perfectly) and others even crash the entire application.

I will post details in other post, plus this happens in older plugin versions too.

**Comment by jepiqueau** [comment]
@PhantomPainX hard to make this working. 
- i create a new project from ionic
- i generate a new page videos and have included your code 
- i add videos.page.html
 ```
<ion-content>
  <!-- Mandatory id="fullscreen" -->
  <div id="fullscreen" slot="fixed">
  </div>

</ion-content>

 ```
- i modify the home.page.html 

```
  <div id="container">
    <ion-button (click)="goVideosPage()" color='primary' expand="block">Videos</ion-button>
  </div>
```
- modify the home.page.ts

```
import { Component, OnInit } from '@angular/core';
import { ModalController } from '@ionic/angular';
import { Capacitor } from '@capacitor/core';
import { VideosPage } from '../videos/videos.page';

@Component({
  selector: 'app-home',
  templateUrl: 'home.page.html',
  styleUrls: ['home.page.scss'],
})
export class HomePage implements OnInit {

  constructor(public modalCtrl: ModalController) {}

  ngOnInit(): void {
    const platform = Capacitor.getPlatform();
    console.log(`$$$ platform: ${platform}`);
  }
  async goVideosPage() {
    const modal = await this.modalCtrl.create({
      component: VideosPage,
      swipeToClose: true
    });
    await modal.present();
  }

}
```
and i got the following error in iOS
```
[error] - ERROR Error: Uncaught (in promise): NullInjectorError: R3InjectorError(HomePageModule)[HTTP -> HTTP -> HTTP -> HTTP]: 
  NullInjectorError: No provider for HTTP!

```
Do i have to add something in home.module.ts or in videos.module.ts ?

**Comment by PhantomPainX** [comment]
> @PhantomPainX hard to make this working. 
> - i create a new project from ionic
> - i generate a new page videos and have included your code 
> - i add videos.page.html
>  ```
> <ion-content>
>   
>   <div id="fullscreen" slot="fixed">
>   </div>
> 
> </ion-content>
> 
>  ```
> - i modify the home.page.html 
> 
> ```
>   <div id="container">
>     <ion-button (click)="goVideosPage()" color='primary' expand="block">Videos</ion-button>
>   </div>
> ```
> - modify the home.page.ts
> 
> ```
> import { Component, OnInit } from '@angular/core';
> import { ModalController } from '@ionic/angular';
> import { Capacitor } from '@capacitor/core';
> import { VideosPage } from '../videos/videos.page';
> 
> @Component({
>   selector: 'app-home',
>   templateUrl: 'home.page.html',
>   styleUrls: ['home.page.scss'],
> })
> export class HomePage implements OnInit {
> 
>   constructor(public modalCtrl: ModalController) {}
> 
>   ngOnInit(): void {
>     const platform = Capacitor.getPlatform();
>     console.log(`$$$ platform: ${platform}`);
>   }
>   async goVideosPage() {
>     const modal = await this.modalCtrl.create({
>       component: VideosPage,
>       swipeToClose: true
>     });
>     await modal.present();
>   }
> 
> }
> ```
> and i got the following error in iOS
> ```
> [error] - ERROR Error: Uncaught (in promise): NullInjectorError: R3InjectorError(HomePageModule)[HTTP -> HTTP -> HTTP -> HTTP]: 
>   NullInjectorError: No provider for HTTP!
> 
> ```
> Do i have to add something in home.module.ts or in videos.module.ts ?
> 

You have to install Cordova http plugin (follow steps in the plugin page) and add the http module in providers array in app.module.ts

**Comment by jepiqueau** [comment]
@PhantomPainX this is what i did now and it is working fine in iOS showing the first vidéo but nt working on Android do no show the player with no error on the Android console

**Comment by PhantomPainX** [comment]
@jepiqueau are you using your last version v4.0.1? The version I tested it was with V3.7.3

**Comment by jepiqueau** [comment]
@PhantomPainX yes i do the Android code did not change

**Comment by PhantomPainX** [comment]
> @PhantomPainX yes i do the Android code did not change 

Tell me if I can do something to help you, without any error shown is difficult

**Comment by jepiqueau** [comment]
@PhantomPainX i fix the bug in 4.1.0-1 it was related to the permissions added for reading internal files. so do not use 3.7.4 which is also buggy.
The video is well played on android but the cast to my TV does not work. so i have to trust you . i will publish the testvideo project i use to test if you can confirm that it works it your side [testvideo](https://github.com/jepiqueau/testvideo.git)

**Comment by PhantomPainX** [comment]
> @PhantomPainX i fix the bug in 4.1.0-1 it was related to the permissions added for reading internal files. so do not use 3.7.4 which is also buggy.
> The video is well played on android but the cast to my TV does not work. so i have to trust you . i will publish the testvideo project i use to test if you can confirm that it works it your side [testvideo](https://github.com/jepiqueau/testvideo.git)

Sure, I will test it once I'm free

**Comment by PhantomPainX** [comment]
@jepiqueau I tested the 4.1.0-1 version and cast works

![Screenshot_2022-09-06-16-19-22-85_21ae4d291e4b2391d104fa7a6e40562e.jpg](https://user-images.githubusercontent.com/47803967/188731393-8fc1cd8c-53a6-4bb5-a8cd-d1fdb497d8de.jpg)
![Screenshot_2022-09-06-16-19-27-20_21ae4d291e4b2391d104fa7a6e40562e.jpg](https://user-images.githubusercontent.com/47803967/188731411-2b228fe9-1c59-4bc9-a08a-324cc74d386a.jpg)
![Screenshot_2022-09-06-16-19-55-29_21ae4d291e4b2391d104fa7a6e40562e.jpg](https://user-images.githubusercontent.com/47803967/188731426-2d940315-d993-451d-aaab-0ea201a1a61c.jpg)

**Comment by jepiqueau** [comment]
@PhantomPainX thanks for your feeback. Do you have a special device connected to your TV. Mine is chromecast enable but it does not work it is a pity

**Comment by PhantomPainX** [comment]
> @PhantomPainX thanks for your feeback. Do you have a special device connected to your TV. Mine is chromecast enable but it does not work it is a pity

I have a Chromecast 3rd generation connected to my TV and a Google nest mini

**Comment by jepiqueau** [comment]
@PhantomPainX i will close this one as everything works as expected Thanks again for your contribution
